File: ios/Ninehoto/Utilities/DeepLinkHandler.swift
import Foundation

enum DeepLink: String {
    case home
    case settings
    case statistics
    case about
    
    case swipeSession
    case finishConfirmation
    
    case unknown
    
    init(url: URL) {
        guard let host = url.host else {
            self = .unknown
            return
        }
        
        switch host {
        case "home": self = .home
        case "settings": self = .settings
        case "statistics": self = .statistics
        case "about": self = .about
        case "session": self = .swipeSession
        case "finish": self = .finishConfirmation
        default: self = .unknown
        }
    }
    
    var path: String {
        switch self {
        case .home: return "/"
        case .settings: return "/settings"
        case .statistics: return "/statistics"
        case .about: return "/about"
        case .swipeSession: return "/session"
        case .finishConfirmation: return "/session/finish"
        case .unknown: return ""
        }
    }
}

final class DeepLinkHandler {
    static let shared = DeepLinkHandler()
    
    private init() {}
    
    func handle(_ url: URL) -> Bool {
        let deepLink = DeepLink(url: url)
        
        guard deepLink != .unknown else {
            Logger.shared.warning("Unknown deep link: \(url)")
            return false
        }
        
        Logger.shared.info("Handling deep link: \(deepLink.rawValue)")
        
        switch deepLink {
        case .settings:
            return navigateToSettings()
        case .statistics:
            return navigateToStatistics()
        case .about:
            return navigateToAbout()
        default:
            return true
        }
    }
    
    private func navigateToSettings() -> Bool {
        NotificationCenter.default.post(name: .navigateToSettings, object: nil)
        return true
    }
    
    private func navigateToStatistics() -> Bool {
        NotificationCenter.default.post(name: .navigateToStatistics, object: nil)
        return true
    }
    
    private func navigateToAbout() -> Bool {
        NotificationCenter.default.post(name: .navigateToAbout, object: nil)
        return true
    }
}

extension Notification.Name {
    static let navigateToSettings = Notification.Name("navigateToSettings")
    static let navigateToStatistics = Notification.Name("navigateToStatistics")
    static let navigateToAbout = Notification.Name("navigateToAbout")
}

File: android/Ninehoto.Android/Utilities/DeepLinkHandler.cs
using System;
using Android.Content;

namespace Ninehoto.Android.Utilities
{
    public enum DeepLink
    {
        Home,
        Settings,
        Statistics,
        About,
        SwipeSession,
        FinishConfirmation,
        Unknown
    }

    public class DeepLinkHandler
    {
        private static DeepLinkHandler? _instance;
        public static DeepLinkHandler Instance => _instance ??= new DeepLinkHandler();

        private readonly ILoggingService _logger;

        private DeepLinkHandler()
        {
            _logger = new LoggingService();
        }

        public DeepLink Parse(Android.Net.Uri uri)
        {
            if (uri == null || string.IsNullOrEmpty(uri.Host))
                return DeepLink.Unknown;

            return uri.Host.ToLowerInvariant() switch
            {
                "home" => DeepLink.Home,
                "settings" => DeepLink.Settings,
                "statistics" => DeepLink.Statistics,
                "about" => DeepLink.About,
                "session" => DeepLink.SwipeSession,
                "finish" => DeepLink.FinishConfirmation,
                _ => DeepLink.Unknown
            };
        }

        public bool Handle(Android.Net.Uri uri)
        {
            var deepLink = Parse(uri);

            if (deepLink == DeepLink.Unknown)
            {
                _logger.Warning($"Unknown deep link: {uri}");
                return false;
            }

            _logger.Info($"Handling deep link: {deepLink}");

            switch (deepLink)
            {
                case DeepLink.Settings:
                    NavigateToSettings();
                    return true;
                case DeepLink.Statistics:
                    NavigateToStatistics();
                    return true;
                case DeepLink.About:
                    NavigateToAbout();
                    return true;
                default:
                    return true;
            }
        }

        private void NavigateToSettings()
        {
            var intent = new Intent(Android.App.Application.Context, typeof(SettingsActivity));
            intent.AddFlags(ActivityFlags.NewTask);
            Android.App.Application.Context.StartActivity(intent);
        }

        private void NavigateToStatistics()
        {
            var intent = new Intent(Android.App.Application.Context, typeof(StatisticsActivity));
            intent.AddFlags(ActivityFlags.NewTask);
            Android.App.Application.Context.StartActivity(intent);
        }

        private void NavigateToAbout()
        {
            var intent = new Intent(Android.App.Application.Context, typeof(AboutActivity));
            intent.AddFlags(ActivityFlags.NewTask);
            Android.App.Application.Context.StartActivity(intent);
        }

        public string GetPath(DeepLink link)
        {
            return link switch
            {
                DeepLink.Home => "/",
                DeepLink.Settings => "/settings",
                DeepLink.Statistics => "/statistics",
                DeepLink.About => "/about",
                DeepLink.SwipeSession => "/session",
                DeepLink.FinishConfirmation => "/session/finish",
                DeepLink.Unknown => "",
                _ => ""
            };
        }
    }
}